### PR TITLE
Hamiltonian cycle implementation

### PR DESCRIPTION
**Pull request by itsAkshayDubey**
### **Describe your change:**

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

#### References

https://github.com/TheAlgorithms/Java/issues/3057 Implementation of Hamiltonian Cycle

### **Checklist:**

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new Java files are placed inside an existing directory.
- [x] All filenames are in all uppercase characters with no spaces or dashes.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: src/main/java/com/thealgorithms/datastructures/graphs/HamiltonianCycle.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	/**
+	 * Find paths recursively from given vertex
+	 *
+	 * @param vertex Vertex from which path is to be found
+	 */
+	public void solve(int vertex) throws Exception{
+
+		if (this.graph[vertex][0] == 1 && this.pathCount == this.V) {
+
+			throw new Exception("Solution found");
```
Comment (siriak) [inline src/main/java/com/thealgorithms/datastructures/graphs/HamiltonianCycle.java:60]: Please don't use exceptions for flow control. return found path from the function instead

**Review by siriak — CHANGES_REQUESTED** [review]
Please also format your code to use spaces instead of tabs and not to contain so many empty lines

File: src/main/java/com/thealgorithms/datastructures/graphs/HamiltonianCycle.java (hunk 1 above)
Comment (itsAkshayDubey) [reply src/main/java/com/thealgorithms/datastructures/graphs/HamiltonianCycle.java]: Hello @siriak 

Formatted code in commit: 57d46aa5c1896209afe45875bd2670b33ba1d1dd, fixed flow of solve method and renamed + changed return type of _void solve(int x)_ to  _boolean isPathFound(int x)_ in commit: 7c343a92fc21921561a6b6f4cdcd4af5ed5daeb5

Kindly have a look.

Thanks.

File: src/test/java/com/thealgorithms/datastructures/graphs/HamiltonianCycleTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.Test;
+
+class HamiltonianCycleTest {
+
+	private HamiltonianCycle hamiltonianCycle = new HamiltonianCycle();
+
+	@Test
+	void testFindHamiltonianCycleShouldReturnHamiltonianPath() {
```
Comment (siriak) [inline src/test/java/com/thealgorithms/datastructures/graphs/HamiltonianCycleTest.java:12]: If it's called `findCycle`, it should return a cycle, not just a path. Please rename everything to `find path`, not `find cycle` if that's what it's supposed to do

**Review by siriak — CHANGES_REQUESTED** [review]

**Review by itsAkshayDubey — COMMENTED** [review]
refactor: Change test case name

**Review by siriak — COMMENTED** [review]
Could you fix the build?

**Comment by itsAkshayDubey** [comment]
Closing this PR - due to some issue with this remote branch I am getting internal server when trying to push the changes.

[Link ](https://github.com/github-community/community/discussions/18940)to issue raised in github community.

Will create new branch and create a new PR with all the suggested changes.

Thanks

**Pull request closed without merging**